### PR TITLE
Add interval functions add/subtract months, seconds etc.

### PR DESCRIPTION
**Pull request by junli1026**
I hereby agree to the terms of the CLA available at: https://databend.rs/policies/cla/

## Summary

Add functions of add/subtract years, months, days, hours, minutes and seconds

## Changelog

- New Feature

## Test Plan

Unit Tests

Stateless Test

### CODE REVIEW

File: common/functions/src/scalars/dates/interval_function.rs [hunk 1]
```diff
@@ -0,0 +1,727 @@
...
+        let (_y, _m, d, h, m, s) = (
+            dt.year(),
+            dt.month(),
+            dt.day(),
+            dt.hour(),
+            dt.minute(),
+            dt.second(),
+        );
+
+        // Handle month last day overflow, "2020-2-29" + "1 year" should be "2020-2-28", or "1990-1-31" + "3 month" should be "1990-4-30".
```
Comment (sundy-li) [inline common/functions/src/scalars/dates/interval_function.rs:501]:
```suggestion
        // Handle month last day overflow, "2020-2-29" + "1 year" should be "2021-2-28", or "1990-1-31" + "3 month" should be "1990-4-30".
```

File: common/functions/src/scalars/dates/interval_function.rs [hunk 2]
```diff
@@ -0,0 +1,727 @@
...
+        let (interval, date) = Self::validate_input(op, a, b)?;
+        let milliseconds_per_day = 24 * 3600 * 1000;
+
+        let days: DataColumn = interval
+            .column()
+            .to_array()?
+            .i64()?
+            .apply(|ms| ms / milliseconds_per_day)
+            .into();
+        date.column().arithmetic(op.clone(), &days)
```
Comment (sundy-li) [inline common/functions/src/scalars/dates/interval_function.rs:256]: If `date` is `Date16 or Date32`, the result of `date.column().arithmetic` will be  `i64` column, and it represents days.

**Comment by junli1026** [comment]
Ping :)

**Comment by sundy-li** [comment]
@junli1026  
Have you read my comment message in `interval_daytime_plus_minus_date`?
 
 I think it's wrong to apply `column.arithmetic`, because `date16` minus `i64` will be `i64`.

**Comment by sundy-li** [comment]
Can you add stateless-tests  SQL tests for this pr?

**Comment by junli1026** [comment]
> Can you add stateless-tests SQL tests for this pr?

sure will do.

**Comment by junli1026** [comment]
> @junli1026
> Have you read my comment message in `interval_daytime_plus_minus_date`?
> 
> I think it's wrong to apply `column.arithmetic`, because `date16` minus `i64` will be `i64`.

Yes, we have some logic to convert i64 to date16 after the calculation. Are you worried about the overflow ?

**Comment by sundy-li** [comment]
> Yes, we have some logic to convert i64 to date16 after the calculation

OK, seems it's ok. But `result.cast_with_type(&data_type)` need extra cast operation on the array. I'd like to keep it in a single loop for better performance.

Can we make op as an argument as arrow2 did ?
```
#[inline]
pub fn binary<T, D, F>(
    lhs: &PrimitiveArray<T>,
    rhs: &PrimitiveArray<D>,
    data_type: DataType,
    op: F,
) -> Result<PrimitiveArray<T>>
where
    T: NativeType,
    D: NativeType,
    F: Fn(T, D) -> T,
```

```
Date32 op interval --> binary(lhs, rhs,  DataType::UInt32,   |a, b|   (a as u32 + b/mills   as u32)   )
Date16 op interval --> binary(lhs, rhs,  DataType::UInt16,   |a, b|   (a as u162 + b/mills    as u16)   )

```

**Review by sundy-li — APPROVED** [review]
Now looks great to me, thanks @junli1026

**Comment by junli1026** [comment]
Will add doc in another PR soon.

**Pull request merged**